File: model_zoo/vision/ppseg/ppseg_unet.py
import fastdeploy as fd
import cv2
import tarfile

# 下载模型和测试图片
model_url = "https://github.com/felixhjh/Fastdeploy-Models/raw/main/unet_Cityscapes.tar.gz"
test_jpg_url = "https://paddleseg.bj.bcebos.com/dygraph/demo/cityscapes_demo.png"
fd.download(model_url, ".", show_progress=True)
fd.download(test_jpg_url, ".", show_progress=True)

try:
    tar = tarfile.open("unet_Cityscapes.tar.gz", "r:gz")
    file_names = tar.getnames()
    for file_name in file_names:
        tar.extract(file_name, ".")
    tar.close()
except Exception as e:
    raise Exception(e)

# 加载模型
model = fd.vision.ppseg.Model("./unet_Cityscapes/model.pdmodel",
                              "./unet_Cityscapes/model.pdiparams",
                              "./unet_Cityscapes/deploy.yaml")

# 预测图片
im = cv2.imread("./cityscapes_demo.png")
result = model.predict(im)

vis_im = im.copy()
# 可视化结果
fd.vision.visualize.vis_segmentation(im, result, vis_im)
cv2.imwrite("vis_img.png", vis_im)

# 输出预测结果
print(result)
print(model.runtime_option)

File: examples/vision/ppseg_unet.cc
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "fastdeploy/vision.h"
#include "yaml-cpp/yaml.h"

int main() {
  namespace vis = fastdeploy::vision;

  std::string model_file = "../resources/models/unet_Cityscapes/model.pdmodel";
  std::string params_file =
      "../resources/models/unet_Cityscapes/model.pdiparams";
  std::string config_file = "../resources/models/unet_Cityscapes/deploy.yaml";
  std::string img_path = "../resources/images/cityscapes_demo.png";
  std::string vis_path = "../resources/outputs/vis.jpeg";

  auto model = vis::ppseg::Model(model_file, params_file, config_file);
  if (!model.Initialized()) {
    std::cerr << "Init Failed." << std::endl;
    return -1;
  }

  cv::Mat im = cv::imread(img_path);
  cv::Mat vis_im;

  vis::SegmentationResult res;
  if (!model.Predict(&im, &res)) {
    std::cerr << "Prediction Failed." << std::endl;
    return -1;
  } else {
    std::cout << "Prediction Done!" << std::endl;
  }

  // 输出预测框结果
  std::cout << res.Str() << std::endl;

  YAML::Node cfg = YAML::LoadFile(config_file);
  int num_classes = 19;
  if (cfg["Deploy"]["num_classes"]) {
    num_classes = cfg["Deploy"]["num_classes"].as<int>();
  }

  // 可视化预测结果
  vis::Visualize::VisSegmentation(im, res, &vis_im, num_classes);
  cv::imwrite(vis_path, vis_im);
  std::cout << "Inference Done! Saved: " << vis_path << std::endl;
  return 0;
}
